sw/device/lib/crypto/drivers/aes.c

```diff
@@ -68,23 +68,36 @@ static status_t aes_write_key(aes_key_t key) {
   // Handle key shares in two separate loops to avoid dealing with
   // corresponding parts too close together, which could risk power
   // side-channel leakage in the ALU.
-  // TODO: randomize iteration order.
-  size_t i = 0;
-  for (; i < key.key_len; ++i) {
+  // Start from a random index less than `key.key_len`.
+  size_t i = ((uint64_t)ibex_rnd32_read() * key.key_len) >> 32;
+  enum { kStep = 1 };
+  size_t iter_cnt = 0, r_iter_cnt = key.key_len - 1;
+  for (; launder32(iter_cnt) < key.key_len;
+       iter_cnt = launderw(iter_cnt) + 1,
+       r_iter_cnt = launderw(r_iter_cnt) - 1) {
     abs_mmio_write32(share0 + i * sizeof(uint32_t), key.key_shares[0][i]);
+    i = (i + kStep) % key.key_len;
```

Comment:
this should in many cases translate to a remu instr

@h-filali please lets add a comment re constant time and what we expect here

Reply:
Is there a way to avoid the REMU here? Because it will take 37 clock cycles. My guess is that only for 192-bit keys, the REMU is inserted.

Can you please confirm that key_shares[*][*] points to an array that is always 8 32-bit values long? If yes, this whole code can probably be optimized (both performance and hardening wise). Right now, the driver writes the used key values in a randomized order to the HW, then it writes randomized values down to the unused words in the hardware. Instead, I suggest to first fill the unused words with random data in memory, and then write the 16 times 32-bit values in randomized order down to the hardware. Meaning you don't need REMU and you randomize between actual key material and unused data.